staging/src/k8s.io/apiserver/pkg/util/flowcontrol/filterconfig/apf_controller.go

```diff
@@ -0,0 +1,700 @@
+
+	// Supply missing mandatory FlowSchemas, in correct position
+	if !meal.haveExemptFS {
+		fsSeq = append(apihelpers.FlowSchemaSequence{fcboot.MandatoryFlowSchemaExempt}, fsSeq...)
+	}
+	if !meal.haveCatchAllFS {
+		fsSeq = append(fsSeq, fcboot.MandatoryFlowSchemaCatchAll)
+	}
+
+	meal.cfgCtl.flowSchemas = fsSeq
+	if klog.V(5) {
+		for _, fs := range fsSeq {
+			klog.Infof("Using FlowSchema %#+v", fcfmt.Fmt(fs))
+		}
+	}
+}
+
+// Consider all the priority levels in the previous configuration.
+// Keep the ones that are in the new config, supply mandatory
+// behavior, or are still busy; for the rest: drop it if it has no
+// queues, otherwise start the quiescing process if that has not
+// already been started.
+func (meal *cfgMeal) processOldPLsLocked() {
+	for plName, plState := range meal.cfgCtl.priorityLevelStates {
+		if meal.newPLStates[plName] != nil {
+			// Still desired and already updated
+			continue
+		}
+		if plName == fctypesv1a1.PriorityLevelConfigurationNameExempt && !meal.haveExemptPL || plName == fctypesv1a1.PriorityLevelConfigurationNameCatchAll && !meal.haveCatchAllPL {
+			klog.V(3).Infof("Retaining mandatory priority level %q despite lack of API object", plName)
+		} else {
+			if plState.queues == nil || plState.quiescing && plState.numPending == 0 && plState.queues.IsIdle() {
+				// Either there are no queues or they are done
+				// draining and no use is coming from another
+				// goroutine
+				klog.V(3).Infof("Removing undesired priority level %q, Type=%v", plName, plState.config.Type)
+				continue
+			}
+			if !plState.quiescing {
+				klog.V(3).Infof("Priority level %q became undesired", plName)
+				plState.quiescing = true
+			}
+		}
+		var err error
+		plState.qsCompleter, err = qscOfPL(meal.cfgCtl.queueSetFactory, plState.queues, plName, &plState.config, meal.cfgCtl.requestWaitLimit)
+		if err != nil {
+			// This can not happen because qscOfPL already approved this config
+			panic(fmt.Sprintf("%s from name=%q spec=%#+v", err.Error(), plName, fcfmt.Fmt(plState.config)))
+		}
+		if plState.config.Limited != nil {
+			meal.shareSum += float64(plState.config.Limited.AssuredConcurrencyShares)
+		}
+		meal.haveExemptPL = meal.haveExemptPL || plName == fctypesv1a1.PriorityLevelConfigurationNameExempt
+		meal.haveCatchAllPL = meal.haveCatchAllPL || plName == fctypesv1a1.PriorityLevelConfigurationNameCatchAll
+		meal.newPLStates[plName] = plState
+	}
+}
+
+// For all the priority levels of the new config, divide up the
+// server's total concurrency limit among them and create/update their
+// QueueSets.
+func (meal *cfgMeal) finishQueueSetReconfigsLocked() {
+	for plName, plState := range meal.newPLStates {
+		if plState.config.Limited == nil {
+			klog.V(5).Infof("Using exempt priority level %q: quiescing=%v", plName, plState.quiescing)
+			continue
+		}
+
+		concurrencyLimit := int(math.Ceil(float64(meal.cfgCtl.serverConcurrencyLimit) * float64(plState.config.Limited.AssuredConcurrencyShares) / meal.shareSum))
```

Comment:
nit: better if we comment here that applying ceiling can lead to some inaccuracy but it's by-design ignored in the current phase. otherwise future-visitors can be confused..

Reply:
done